include/deal.II/numerics/smoothness_estimator.h

```diff
@@ -0,0 +1,268 @@
+{
+  /**
+   * Estimate smoothness from decay of Legendre absolute values of coefficients
+   * on the reference cell.
+   *
+   * In one dimension, the finite element solution on the reference element with
+   * polynomial degree $p$ can be written as
+   * @f[
+   *    u(x) = \sum_{j=0}^{p} a_j P_j(x)
+   * @f]
+   * where $\{P_j(x)\}$ are Legendre polynomials. The decay of the coefficients
+   * is estimated by performing the least squares fit of
+   * @f[
+   *   |a_j| \sim C exp(-\sigma j)
+   * @f]
+   * for $j=1,..,p$. The rate of the decay can be used to estimate the
+   * smoothness. For example, p-refinement will be performed if $\sigma>1$.
+   *
+   * Extension to higher dimension is done by performing the fit in each
+   * direction separately and then taking the lowest value of $\sigma$.
+   *
+   * The returned value for each vector and for each cell is $\exp(-\sigma)$.
+   *
+   * @param fe_series FESeries::Legendre object to calculate coefficients. This
+   * object needs to be initialized to have at least $p+1$ coefficients in each
+   * direction, where $p$ is the maximum polynomial degree to be used.
+   * @param dof_hander an hp::DoFHandler
+   * @param all_solutions a vector of pointers to the solution vectors
+   * @param all_smoothness_indicators a vector of pointers to the smoothness indicators for each @p all_solutions.
+   * @param smallest_abs_coefficient is the smallest absolute value of the
+   * coefficient to be used in linear regression. In case there are less than
+   * two non-zero coefficients, the direction will be skipped. If all
+   * coefficients are zero, the returned value for this cell will be zero (i.e.
+   * smooth).
+   *
+   * @note For linear elements the returned value is always zero.
+   *
+   * For more theoretical details see
+   * @code{.bib}
+   * @Article{Mavriplis1994,
+   *  author    = {Mavriplis, Catherine},
+   *  title     = {Adaptive mesh strategies for the spectral element method},
+   *  journal   = {Computer methods in applied mechanics and engineering},
+   *  year      = {1994},
+   *  volume    = {116},
+   *  number    = {1},
+   *  pages     = {77--86},
+   *  publisher = {Elsevier},
+   * }
+   * @endcode
+   * and for the application within the deal.II:
+   * @code{.bib}
+   * @article{Davydov2017,
+   *   author  = {Denis Davydov and Tymofiy Gerasimov and Jean-Paul Pelteret and
+   *              Paul Steinmann},
+   *   title   = {Convergence study of the h-adaptive PUM and the hp-adaptive
+   * FEM applied to eigenvalue problems in quantum mechanics},
+   *   journal = {Advanced Modeling and Simulation in Engineering Sciences},
+   *   year    = {2017},
+   *   volume  = {4},
+   *   number  = {1},
+   *   pages   = {7},
+   *   issn    = {2213-7467},
+   *   doi     = {10.1186/s40323-017-0093-0},
+   * }
+   * @endcode
+   */
+  template <int dim, int spacedim, typename VectorType>
+  void
+  legendre_coeff_decay(
+    FESeries::Legendre<dim, spacedim> &    fe_series,
```

Comment:
The description of the argument above makes it sound like this is an input argument (because you say that it is initialized with something). But it's not `const`. Can you explain why this is so?

Reply:
it needs to use this object to do the job of calculating coefficients, it's methods to do so aren't `const`, that's why it's not `const` either.

Reply:
I see. Specifically which functions are you calling that aren't `const`?

Reply:
The `calculate` function also fills the transform matrices on demand in private class member variables. This will be performed by the `ensure_existence` functions which are non-`const`.

Reply:
I see, thanks!

(Variables that are cached could also be made `mutable` and guarded by a mutex. But that's a question separate from what we discuss here.)